### ISSUE TITLE
Hiding column makes unwanted padding... 

### ISSUE THREAD

**Issue by MikhailKlemin** [olifolkerd/tabulator#139]
Hello! 
There is a grey space on the right. 
![screenshot from 2016-12-18 22-56-12](https://cloud.githubusercontent.com/assets/8426672/21296142/f4c3a292-c57d-11e6-99e9-990904fb0b5a.png)

![screenshot from 2016-12-18 22-56-31](https://cloud.githubusercontent.com/assets/8426672/21296143/fa29b88e-c57d-11e6-867d-6aabff491599.png)

![screenshot from 2016-12-18 23-01-19](https://cloud.githubusercontent.com/assets/8426672/21296158/4c4983ec-c57e-11e6-8872-ea322bc1206e.png)

**Comment by olifolkerd** [comment]
Could you post a copy of your table constructor and a few lines of your test data and i will find out whats up.

How are you hiding the column, are you using the ***visible*** option in the column definition or are you using the ***hideCol*** function?

Cheers

Oli

**Comment by MikhailKlemin** [comment]
Hello!
This is constructor, I am sorry it's a little messed up, as I am trying various things.
Also worth to mention, that it  takes time to digest Ajax even very small batch of data, takes table to "think" for a second or two. 
I use hideColl function as you can see, right after constructor.

I never used JavaScript before a month I think, so might be some silly mistakes on my side ;-( 

```
$(document).ready(function() {
	var filterIcon = function(value, data, cell, row, operations){
		return "<i class='fa fa-minus' aria-hidden='true'></i>";
	};
   
    $("#example-table").tabulator({
        progressiveRenderSize: 10,
        fitColumns:true,
        pagination:false,
        sortBy:'Days', // when data is loaded into the table, sort it by name
        sorDir:'desc', //when data is loaded into the table, sort it in ascending order.
        index: "ID",
        showLoader: true,
        paginationSize:50,
        columns:[ //Define Table Columns
 {title:"Days", field:"Days", sorter:"number", width:75, align:"left" },
 {title:"Orig date", field:"TransactionDate",sorter:"date", width:150, align:"left" },
 {title:"Last Activity", field:"ModifiedAfter", sorter:"date", width:150, align:"left" },
 {title:"Category", field:"CategoryText", sorter:"string", width:150, sortable:false }, 
 {title:"Desc", field:"Desc", sorter:"string", width:300, sortable:false },
 {title:"Control", field:"Control", sorter:"string", width:100, align:"left" },
 {title:"Vehicle Cost", field:"VehicleCost", sorter:"number", width:150, align:"left" }, 
 {title:"Account", field:"Account", sorter:"string", width:130, editable:false, },
 {title:"Involice", field:"Invoice", sorter:"string", width:130, editable:false },
 {title:"Curtailment", field:"Curtailment", sorter:"string", width:130, editable:false },
 {title:"Amount", field:"Amount", sorter:"number", width:100, align:"left"},           
 {title:"<div class='text-center'><i class='fa fa-filter' aria-hidden='true'></i></div>", field:"Filtered", formatter:filterIcon, width:50, sortable:false,   align:"center", 
                onClick:function(e, cell, value, data){
                    //hidden
                            });
                        }      
                    }
            },
            {title:"Comment",   field:"Comment",sorter:"string",   align:"left",   editable:true}
        ],
        cellEdited:function(id,data,row){
          //hidden
            }); 
       },

    });
    $("#example-table").tabulator("setData","http://{{.myHost}}/table?catID=" + $("#navtabs li.active").attr("id"));
    $("#example-table").tabulator("hideCol","Invoice");
    $("#example-table").tabulator("hideCol","Curtailment");    
    $("#example-table").tabulator("hideCol","TransactionDate");
    $("#example-table").tabulator("hideCol","ModifiedAfter");
    $("#example-table").tabulator("hideCol","VehicleCost");
    
});
```

When user navigates menu, there is a function called like:
```
$('#navtabs li').on("mouseup",function(e){
    console.log(e.currentTarget.id);
    $('#navtabs li').removeClass("active");
    $("#"+e.currentTarget.id).addClass("active");   
    if (e.currentTarget.id == {{.FloorPlanID}}){
    	$("#example-table").tabulator("showCol","Invoice");
    	$("#example-table").tabulator("showCol","Curtailment");
        $("#example-table").tabulator("showCol","VehicleCost");
    }else{
    	$("#example-table").tabulator("hideCol","Invoice");
    	$("#example-table").tabulator("hideCol","Curtailment");
        $("#example-table").tabulator("hideCol","VehicleCost");
    }

    if (e.currentTarget.id == 1976){
        $("#example-table").tabulator("showCol","TransactionDate");
        $("#example-table").tabulator("showCol","ModifiedAfter");
    }else{
        $("#example-table").tabulator("hideCol","TransactionDate");
        $("#example-table").tabulator("hideCol","ModifiedAfter");
    }
    $("#example-table").tabulator("setData", "http://{{.myHost}}/table" + "?catID=" + e.currentTarget.id);
    return false;
});
```

**Comment by olifolkerd** [comment]
Having looked at your pictures again, that grey space is supposed to be there, that is the background colour of the table, it displays when the columns aren't wide enough to fill the table.

The issue you are having is that by default Tabulator will not alter the widths of the remaining columns when a column is show or hidden. to get round this you can call the ***redraw*** function after you have hidden/shown columns to cause Tabulator to recalculate the widths to fill the table, eg:

```js
$("#example-table").tabulator("showCol","Invoice");
$("#example-table").tabulator("showCol","Curtailment");
$("#example-table").tabulator("showCol","VehicleCost");
$("#example-table").tabulator("redraw"); //Redraw columns to fit the table width
```

In reguards to the loading, if you are loading data via ajax, it will take a second or so to process the data because it has to make the request to the server, wait for a response, then process the data. If you are loading a lot of rows then I would suggest using the [Progressive Rendering](http://olifolkerd.github.io/tabulator/docs/#progressive-render) feature, which will speed things up.

Let me know if that has solved your issue.

Cheers

Oli

**Comment by MikhailKlemin** [comment]
Hi! Yes, redraw solved problems!
As for Ajax. The Firebug says it takes 1 to 5 ms to perform the request (I am not sure how is it accurate though), but it takes seconds to get table appear, and sometimes I get that warning box (JavaScript takes too long, stop/continue).

**Comment by olifolkerd** [comment]
The processing issue might be arising becaue you have set ***fitColumns:true*** and have also specified a width in almost all your columns so Tabulator is having to work hard to recalculate the widths, you could try removing the width from a few of the columns and see if that helps.

If that doesn't then could you send over the data set that is taking so long to process and i can take a look at it, there may be a data type that Tabulator is struggling with.

What browser (and version) are you working on? have you tried other browsers and do you get the same issue?

Cheers

Oli

**Comment by MikhailKlemin** [comment]
Hi!
I tried either Chrome ( 55.0.2883.87)  and Firefox (50.1.0). 
Here I attach a quick make up  example, it uses JQuery.grep to filter array of objects and then pass it to SetData, it takes sometimes a lot of time, sometimes it is fast, I can not see any patterns though... The Grep works very fast, if you click navigation menu (cat0-cat1-... ) categories it will get very slow.. 
[bootstrap.zip](https://github.com/olifolkerd/tabulator/files/661309/bootstrap.zip)

**Comment by olifolkerd** [comment]
Thanks for sending over the file, that was a great help.

So there are two issue here. firstly your data set is rather large, but not to worry that's what progressive rendering is for.

The reason it isn't working in your example is that you havn't defined a ***height*** for the table in the constructor so it just goes on trying to render all the rows in one go.

```js
$("#example-table").tabulator({
        height:500,
        progressiveRender:true,
        progressiveRenderSize: 10,
        ....
});
```

Secondly, you don't need to grep the data tabulator has an inbuilt ***filter*** function that lets you filter data once it is loaded.

I would recommend you load all of your data into the table immediately after the table has been created
```js
$("#example-table").tabulator("setData", tabledata);
```

Incidentally you can also set the default visible state of a column using the ***visible*** property in its column definition array, I would recommend doing this instead of using the ***hideCol*** function for initially setting up the table as every time you call ***hideCol*** the table gets redrawn which is more resource intensive

```js
{title"Name", field:"name", visible:false} // hide a column in its definition object
```

You can then replace your grep function with the inbuilt ***filter*** function, which will be much more efficient as it doesn't have to re-parse in the data every time:
```js
$("#example-table").tabulator("filter", "Category", e.currentTarget.id)
```
Full documentation on filtering can be found [here](http://olifolkerd.github.io/tabulator/docs/#filtering)

Let me know if that has sorted your issues

Cheers

Oli

**Comment by MikhailKlemin** [comment]
Hi Oli! Thanks a lot!  I tried adding height property - way faster now. I will test more tomorrow, already was sleeping but could not resist to try :-)

**Comment by MikhailKlemin** [comment]
Hello!
Yes,works much much better now 👍 Thank you for your input! 
Just there is no filter method, it is setFilter :-) 

Very last issue I have: When I fix the height to say 500, and there are more rows, it have vertical scroll bar which is fine, but there is also appears unnecessary horizontal scroll bar at the bottom. When all records fit into the specified height, there no scroll at the bottom.

**Comment by olifolkerd** [comment]
Hey,

Great to hear! sorry about the mixup with the filter function, i have updated the previous post with the correct information.

The horizontal scroll bar appears because the column widths are set to take up 100% of the table width then the vertical scrollbar appears as the table overflows which then takes up a fraction of the already allocated space.

This is an issue i cam currently working on. As a quick work around until i have this sorted, calling ***redraw*** after the load or filter should fix the issue :)

**Comment by olifolkerd** [comment]
At long last the issue with the horizontal scroll bar on fit to table column mode has been resolved

**Comment by mahmad2504** [comment]
Tabulator is amazing . I wish you had Visa pay option for donation.... Paypal does not work in my country :)

**Comment by mahmad2504** [comment]
I am using it in my Linux CVE Management tool and i have got every feature that i thought of. Amazing and so easy to use. I salute to your excellent programming skills.
